src/schema/search/SearchResolver.ts

```diff
@@ -0,0 +1,142 @@
+import { convertConnectionArgsToGravityArgs } from "lib/helpers"
+import { createPageCursors, pageToCursor } from "schema/fields/pagination"
+import { connectionFromArraySlice } from "graphql-relay"
+import { GraphQLResolveInfo, visit } from "graphql"
+import { ResolverContext } from "types/graphql"
+import { Searchable } from "schema/searchable"
+import { SearchableItem } from "schema/SearchableItem"
+
+export class SearchResolver {
+  private args: { [argName: string]: any }
+  private context: ResolverContext
+  private info: GraphQLResolveInfo
+  private cachedEntityTypesToFetch: string[] | undefined
+
+  constructor(
+    args: { [argName: string]: any },
+    context: ResolverContext,
+    info: GraphQLResolveInfo
+  ) {
+    this.args = args
+    this.context = context
+    this.info = info
+  }
+
+  fetch(searchResultItem) {
+    const { artistLoader, artworkLoader } = this.context
+    const loaderMapping = {
+      Artist: artistLoader,
+      Artwork: artworkLoader,
+    }
+
+    const loader = loaderMapping[searchResultItem.label]
+
+    if (loader) {
+      return loader(searchResultItem.id)
+    }
+  }
+
+  // Fetch the full object if the GraphQL query includes any inline fragments
+  // referencing the search result item's type (like Artist or Artwork)
+  shouldFetchEntityType(entityType: string): boolean {
+    if (!this.cachedEntityTypesToFetch) {
+      const entityTypesToFetch: string[] = []
+
+      visit(this.info.fieldNodes[0], {
+        Field(node) {
+          if (node.name.value === "node") {
+            visit(node, {
+              InlineFragment(node) {
+                if (
+                  node.typeCondition &&
+                  (node.typeCondition.name.value !== Searchable.name &&
+                    node.typeCondition.name.value !== SearchableItem.name)
+                ) {
+                  entityTypesToFetch.push(node.typeCondition.name.value)
+                }
+              },
+              FragmentSpread(_node) {
+                throw new Error(
+                  "Named fragment spreads are currently unsupported for search."
+                )
+              },
+            })
+          }
+        },
+      })
+
+      this.cachedEntityTypesToFetch = entityTypesToFetch
+    }
+
+    return this.cachedEntityTypesToFetch.includes(entityType)
+  }
+
+  processSearchResultItem(searchResultItem) {
+    if (this.shouldFetchEntityType(searchResultItem.label)) {
+      return this.fetch(searchResultItem).then(response => {
+        return {
+          ...response,
+          __typename: searchResultItem.label,
+        }
+      })
+    } else {
+      return Promise.resolve({
+        ...searchResultItem,
+        __typename: "SearchableItem",
+      })
+    }
+  }
+
+  resolve() {
+    const pageOptions = convertConnectionArgsToGravityArgs(this.args)
```

Comment:
This is missing some critical lines from where they were removed below. See https://github.com/artsy/metaphysics/pull/1643/files#diff-f4af094292cb733a374f6ae47ee2a36cL151 for where it was removed (specifically the `delete args.page`)